File: Hydra/Hydra/ViewModels/ChatViewModel.swift
import Foundation

@MainActor
final class ChatViewModel: ObservableObject {
    @Published private(set) var turns: [ChatTurn] = []
    @Published private(set) var isThinking = false
    @Published var pendingPlan: AgentPlan?
    @Published var pendingPlanMessage: String?
    @Published var lastResults: [ActionResult]?
    @Published var error: String?

    /// History sent to the server is capped at the last 20 turns. The
    /// UI keeps the full list so the user can scroll back.
    private let serverHistoryCap = 20

    private let api = APIClient.shared

    func send(_ message: String, contextPreamble: String? = nil) async {
        let trimmed = message.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        turns.append(ChatTurn(role: "user", content: trimmed, plan: nil, results: nil))
        isThinking = true
        error = nil
        defer { isThinking = false }
        let history = Array(turns.suffix(serverHistoryCap))
        // Preamble is composed by the caller from active tab + selection;
        // we attach it only to the outbound message, never to the on-screen
        // turn text — the user shouldn't see their own boilerplate echoed.
        let outbound: String
        if let preamble = contextPreamble, !preamble.isEmpty {
            outbound = "\(preamble)\n\n\(trimmed)"
        } else {
            outbound = trimmed
        }
        let instr = UserDefaults.standard.string(forKey: "aiInstruction")
        let req = ChatRequest(history: history, message: outbound,
                              instruction: (instr?.isEmpty == false) ? instr : nil)
        do {
            let resp = try await api.chat(req)
            let role = resp.type == "plan" ? "assistant_plan" : "assistant_ask"
            turns.append(ChatTurn(role: role, content: resp.message, plan: resp.plan, results: nil))
            if resp.type == "plan" {
                pendingPlan = resp.plan
                pendingPlanMessage = resp.message
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func runPendingPlan() async {
        guard let plan = pendingPlan else { return }
        isThinking = true
        defer { isThinking = false }
        do {
            let resp = try await api.executePlan(plan)
            lastResults = resp.results
            turns.append(ChatTurn(
                role: "system_result",
                content: summary(of: resp.results),
                plan: nil,
                results: resp.results
            ))
            pendingPlan = nil
            pendingPlanMessage = nil
        } catch {
            self.error = error.localizedDescription
        }
    }

    func cancelPendingPlan() {
        pendingPlan = nil
        pendingPlanMessage = nil
    }

    private func summary(of results: [ActionResult]) -> String {
        let ok = results.filter { $0.status == "ok" }.count
        let fail = results.count - ok
        if fail == 0 { return "✓ all \(ok) action(s) completed" }
        return "ran \(results.count) action(s) — \(ok) ok, \(fail) failed"
    }
}

File: android/feature/chat/src/main/kotlin/com/hydra/android/feature/chat/ChatViewModel.kt
package com.hydra.android.feature.chat

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hydra.android.core.data.ChatRepository
import com.hydra.android.core.data.SettingsSource
import com.hydra.android.core.model.ActionResult
import com.hydra.android.core.model.AgentPlan
import com.hydra.android.core.model.ChatTurn
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ChatUiState(
    val turns: List<ChatTurn> = emptyList(),
    val isThinking: Boolean = false,
    val pendingPlan: AgentPlan? = null,
    val pendingPlanMessage: String? = null,
    val error: String? = null,
)

@HiltViewModel
class ChatViewModel @Inject constructor(
    private val repository: ChatRepository,
    private val settings: SettingsSource,
) : ViewModel() {

    private val _state = MutableStateFlow(ChatUiState())
    val state: StateFlow<ChatUiState> = _state.asStateFlow()

    fun send(message: String) {
        val trimmed = message.trim()
        if (trimmed.isEmpty()) return

        _state.update {
            it.copy(
                turns = it.turns + ChatTurn(role = "user", content = trimmed),
                isThinking = true,
                error = null,
            )
        }

        viewModelScope.launch {
            // The repository caps the outbound history at 20; the UI keeps all
            // of it so the user can still scroll back.
            val history = _state.value.turns
            val instruction = settings.aiInstruction.first()

            repository.send(history, trimmed, instruction).fold(
                onSuccess = { response ->
                    val isPlan = response.type == "plan"
                    _state.update {
                        it.copy(
                            turns = it.turns + ChatTurn(
                                role = if (isPlan) "assistant_plan" else "assistant_ask",
                                content = response.message,
                                plan = response.plan,
                            ),
                            isThinking = false,
                            pendingPlan = if (isPlan) response.plan else null,
                            pendingPlanMessage = if (isPlan) response.message else null,
                        )
                    }
                },
                onFailure = { e ->
                    // The user turn stays in history — losing what they typed
                    // because the network failed would be worse than the error.
                    _state.update { it.copy(isThinking = false, error = e.message) }
                },
            )
        }
    }

    fun runPendingPlan() {
        val plan = _state.value.pendingPlan ?: return
        _state.update { it.copy(isThinking = true, error = null) }

        viewModelScope.launch {
            repository.execute(plan).fold(
                onSuccess = { response ->
                    _state.update {
                        it.copy(
                            turns = it.turns + ChatTurn(
                                role = "system_result",
                                content = summarize(response.results),
                                results = response.results,
                            ),
                            isThinking = false,
                            pendingPlan = null,
                            pendingPlanMessage = null,
                        )
                    }
                },
                onFailure = { e ->
                    _state.update { it.copy(isThinking = false, error = e.message) }
                },
            )
        }
    }

    fun cancelPendingPlan() {
        _state.update { it.copy(pendingPlan = null, pendingPlanMessage = null) }
    }

    /** Wording matches iOS ChatViewModel.summary(of:) exactly. */
    private fun summarize(results: List<ActionResult>): String {
        val ok = results.count { it.isOk }
        val failed = results.size - ok
        return if (failed == 0) "✓ all $ok action(s) completed"
        else "ran ${results.size} action(s) — $ok ok, $failed failed"
    }
}
